lib/merchant.rb

```diff
@@ -43,17 +41,52 @@ def revenue(date=nil)
         invoice.created_at == date
       end
     end
+
     successful_invoice_items = parsed_successful_invoices.map do |invoice|
       invoice.invoice_items
     end.flatten
-    revenue = successful_invoice_items.reduce(0) do |sum, inv_item|
+    successful_invoice_items.reduce(0) do |sum, inv_item|
       sum += inv_item.revenue
     end
-    if !revenue.zero?
-      revenue
+  end
+
+  def favorite_customer
+    merchant_transactions
+
+    successful_transactions = successful_transactions(merchant_transactions)
+
+    successful_invoices = successful_invoices(successful_transactions)
+
+    merchant_customers = successful_invoices.map do |invoice|
+      invoice.customer
+    end
+    merchant_customers.max_by do |customer|
+      merchant_customers.count(customer)
+    end
+  end
+
+  def customers_with_pending_invoices
+    unsuccessful_invoices
+
+    unsuccessful_invoices.map do |invoice|
+      invoice.customer
     end
   end
 
+  def quantity_successful_items
+    merchant_transactions
+
+    successful_transactions = successful_transactions(merchant_transactions)
+
+    successful_invoices = successful_invoices(successful_transactions)
+
+    successful_invoice_items = successful_invoices.map do |invoice|
```

Comment:
Useless assignment to variable - `successful_invoice_items`. Did you mean `successful_invoices`?<br>Pass `&:items` as an argument to `map` instead of a block.